drivers/mdio/mdio_xilinx_axienet.c

```diff
@@ -0,0 +1,200 @@
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+
+	LOG_INF("Disable MDIO Bus!");
+
+	xilinx_axienet_mdio_write_register(config, XILINX_AXIENET_MDIO_SETUP_REG_OFFSET,
+					   XILINX_AXIENET_MDIO_SETUP_REG_MDIO_DISABLE_MASK);
+}
+
+#define MDIO_TARGET_FREQUENCY_HZ 2500000 /* 2.5 MHz, i.e., max MDIO clock according to IEEE spec */
+#define MDIO_BUSY_WAIT_USEC      1
+#define MDIO_BUSY_WAIT_MAX_USEC  10000
+
+static void mdio_bus_wait_ready(const struct mdio_xilinx_axienet_config *config)
+{
+	int attempts = 0;
+
+	while ((xilinx_axienet_read_mdio_register(config, XILINX_AXIENET_MDIO_CONTROL_REG_OFFSET) &
+		XILINX_AXIENET_MDIO_CONTROL_REG_MASK_READY) == 0) {
+
+		LOG_DBG("Waiting for MDIO to become ready!");
+		k_busy_wait(MDIO_BUSY_WAIT_USEC);
+		attempts += MDIO_BUSY_WAIT_USEC;
+		if (attempts >= MDIO_BUSY_WAIT_MAX_USEC) {
+			LOG_ERR("Exceeded max wait time for busy waiting!");
+			k_panic();
+		}
+	}
+}
+
+static void enable_mdio_bus(const struct mdio_xilinx_axienet_config *config,
+			    struct mdio_xilinx_axienet_data *data)
+{
+	xilinx_axienet_mdio_write_register(config, XILINX_AXIENET_MDIO_SETUP_REG_OFFSET,
+					   XILINX_AXIENET_MDIO_SETUP_REG_MDIO_ENABLE_MASK |
+						   data->clock_divider);
+	mdio_bus_wait_ready(config);
+	data->bus_enabled = true;
+}
+
+static void mdio_xilinx_axienet_bus_enable(const struct device *dev)
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+	struct mdio_xilinx_axienet_data *data = dev->data;
+	uint16_t clock_divider, clock_divider_full;
+
+	if (!config->clock_frequency_hz) {
+		LOG_ERR("No clock frequency specified for ethernet device!");
+		return;
+	}
+
+	/* this might result in a MDIO frequency that is a bit lower than the max frequency */
+	clock_divider_full = (config->clock_frequency_hz / (MDIO_TARGET_FREQUENCY_HZ)) * 2;
+	clock_divider = clock_divider_full & XILINX_AXIENET_MDIO_SETUP_REG_MDIO_CLOCK_DIVIDER_MASK;
+
+	if (clock_divider != clock_divider_full) {
+		LOG_ERR("Clock divider overflow!");
+	}
+
+	data->clock_divider = clock_divider;
+
+	LOG_INF("Enable MDIO Bus assuming ethernet clock frequency %u divider %u!",
+		config->clock_frequency_hz, clock_divider);
+
+	xilinx_axienet_mdio_write_register(config, XILINX_AXIENET_MDIO_SETUP_REG_OFFSET,
+					   clock_divider);
+
+	enable_mdio_bus(config, data);
+
+	LOG_INF("MDIO ready!");
+}
+
+static int mdio_xilinx_axienet_read(const struct device *dev, uint8_t prtad, uint8_t devad,
+				    uint16_t *data)
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+	struct mdio_xilinx_axienet_data *dev_data = dev->data;
+
+	enable_mdio_bus(config, dev_data);
+	xilinx_axienet_mdio_write_register(
+		config, XILINX_AXIENET_MDIO_CONTROL_REG_OFFSET,
+		XILINX_AXIENET_MDIO_CONTROL_REG_MASK_INITIATE |
+			(prtad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_PHYADDR) |
+			(devad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_REGADDR) |
+			XILINX_AXIENET_MDIO_CONTROL_REG_MASK_READ);
```

Comment:
just glancing at this, you might be able to make use of the util_macro.h here as well, up to you